### DISCUSSION TITLE
Lightweight alternative to crates.io API and database dumps

### DISCUSSION THREAD

**Discussion by AMDmi3** [rust-lang/crates.io#8454] [General]
Hi, I'm author of https://repology.org, a service which aggregates package repository data and reports new versions to package maintainers. It uses crates.io data as one of its sources, and it's important because more and more end-user software written in Rust is packaged in repositories. However, a problem with getting data from crates.io arised.

Historically, crates data was fetched through API, in a batches of 100, with 1 sec delay between fetches and custom user agent pointing to repology.org (and I think that fulfills [api access requirements](https://crates.io/data-access#api)), however that was broken recently, with crates.io closing connection after a few pages ([example failed fetch log](https://repology.org/log/4277802)). Honestly I wasn't happy with the approach as it took a long time (to fetch 1400+ pages!) and was rather unreliable.

Another option would be to switch to [database dumps](https://static.crates.io/db-dump.tar.gz), but it's too heavy to be processed by repology
- these are generated once a day, while repology strives for delivery times of a few hours at most
- it's much larger, 2.2GB uncompressed vs. 120MB for API output
- getting the same info from it requires joining gigabyte files (at the very least, versions.csv need to be joined with crates.csv)

So, is there a possibility of solution free from disadvantages of these two, which could be a static json dump generated ~hourly and containing the same info which API provides? It would require both less CPU time to generate and process and less traffic to transfer for both crates.io and clients, while providing decent freshness and completeness of the data.

**Comment by carols10cents** [comment]
Can you clarify what data you need, exactly?

**Reply by AMDmi3** [reply]
Currently used API fields are id/name, homepage, repository, documentation, description, max_stable_version and version.

**Comment by carols10cents** [comment]
Also, have you investigated using the index, either [the git repo](https://github.com/rust-lang/crates.io-index) or the [HTTP sparse index](https://doc.rust-lang.org/cargo/reference/registries.html#registry-protocols)?

**Reply by AMDmi3** [reply]
> either [the git repo](https://github.com/rust-lang/crates.io-index)

Still much bigger, lacks links.

> or the [HTTP sparse index](https://doc.rust-lang.org/cargo/reference/registries.html#registry-protocols)?

It doesn't say how to access it without cargo, I suspect it has the same data as git.

**Reply by Turbo87** [reply]
you can use the git repo to watch for changes though and then only request those crates/versions that have been updated in the git repo

**Reply by AMDmi3** [reply]
> you can use the git repo to watch for changes though and then only request those crates/versions that have been updated in the git repo

Unfortunately that's not an option due to complexity.

**Reply by Turbo87** [reply]
https://github.com/rust-lang/crates.io-index/commits.atom ...

**Reply by AMDmi3** [reply]
It requires setting up a whole separate service which watches the feed and fetches info for separate crates with a whole bunch of caveats such as what should it do after going online after a downtime when part of the feed is lost. This requires development and maintenance, I'm definitely not doing it.

**Comment by carols10cents** [comment]
> Honestly I wasn't happy with the approach as it took a long time (to fetch 1400+ pages!) and was rather unreliable.

So one thing I'm not understanding is why do you need to fetch _all_ the crates all the time? You can sort by recent updates https://crates.io/crates?sort=recent-updates and then stop requesting pages when you see updates you've already seen? 1400 * 100 crates are not being updated every hour.

**Reply by AMDmi3** [reply]
Because repology is a batch service and operates on complete snapshots of repository states. As described above, generating such state out of an update stream of updates would require setting up separate stateful service which would be unacceptably complex, as already mentioned above.

**Comment by LawnGnome** [comment]
>This requires development and maintenance, I'm definitely not doing it.

Unfortunately, setting up a new data feed also requires development and maintenance, and given that this data is available in at least two ways (the Git index and database dumps), we don't believe that this complexity is warranted within the crates.io service itself.

docs.rs uses the [`crates-index`](https://crates.io/crates/crates-index) and [`crates-index-diff`](https://crates.io/crates/crates-index-diff) crates to perform this task, and there's also the [`crates-io-cli` command line tool](https://github.com/Byron/crates-io-cli) that uses those to keep a repository up to date and output what's changed since the last update.

**Reply by AMDmi3** [reply]
> data is available in at least two ways

As mentioned, this is not true. git index does not contain all required data, and database dumps are not updated frequently enough.
